### ISSUE TITLE
RDS: Create a restore-only instance from a backup

### ISSUE THREAD

**Issue by flobaader** [pellepelster/solidblocks#18]
If only a subset of data needs to be restored, typically DBaaS services offer a method to create a new cluster from an existing backup.
I could image the same scenario here, where I want to create a new container that restores the state from an old backup but does not write new backups to the backup location.

**Comment by pellepelster** [comment]
release v0.1.18 made first steps in that direction, the Hetzner RDS module now supports multiple start modes (https://pellepelster.github.io/solidblocks/hetzner/rds-postgresql/index.html#input%5c_mode) There seems to be an implementation of the restore-only mode in a private fork, I will see if it is feasible to free the code =)

**Comment by Levara** [comment]
> release v0.1.19 made first steps in that direction, the Hetzner RDS module now supports multiple start modes (https://pellepelster.github.io/solidblocks/hetzner/rds-postgresql/index.html#input%5c_mode) There seems to be an implementation of the restore-only mode in a private fork, I will see if it is feasible to free the code =)

Hi @pellepelster, are there any updates regarding this feature?

**Comment by pellepelster** [comment]
I have a potential working solution but before I dig further into this I would be interested in the use case. Is this for debugging and/or disaster recover scenarios? If so does it have to be a remote instance, or would a local container populated with the data form the backup also be sufficient?

**Comment by Levara** [comment]
Amazing, thanks. The use-case I have in mind is using the same image in both the development and production. How I'd like it to work:
- I have a production database in RDS which backs up to S3
- I have a separate staging database also in RDS that backs up to S3
- I have a local docker compose file that starts the local RDS instance in restore-only mode, pulls the backup but does not perform backups

So when a new developer comes to the project, the only thing they have to do is run the docker-compose file that starts the RDS database and pulls the staging db backup. If they need to update their local db from staging, all they have to do is:
```
docker compose down -v
docker compose up -d
```
And RDS pull the new backup from S3.

Another usecase is supporting per-branch automatic deployments through the CI server. When a developer wants a new branch on the staging server, their code will either have to share a database with branches on the server, or the CI server could spin up a new restore-only instance of the db that would be tied only to that branch. 


>  If so does it have to be a remote instance, or would a local container populated with the data form the backup also be sufficient?

I would be satisfied with the option to simply spin up a new local container that will restore itself from the remote backup, but wont push any backups back to the remote, if that what you're asking.

**Comment by pellepelster** [comment]
how does your setup look like in terms of instance names (https://pellepelster.github.io/solidblocks/hetzner/rds/index.html#input%5c_name) and s3 buckets? 

* does each branch/env have its own instance name and use the same bucket?
* same instance name and different buckets across branches/envs?
* or even a mix of both?

**Comment by Levara** [comment]
For my setup specifically, each project has a separate bucket. Production only has one branch, development (staging) has multiple branches, but only one database, if I deploy additional branch to a different url, it's still pointing at the same database. I have multiple small projects, the majority of db dumps are <5GB uncompressed. 

I currently don't have that "every branch to its own url automatic deployment" thing configured, so I don't have a preference regarding the instance names. I'm not using terraform, I'm starting containers through docker-compose. I'd run all the containers on the same server, just give each container a different hostname on the docker network. Upon reading my post again, I see that I misspoke, I meant "or the CI server could spin up a new restore-only **_container_** of the db ". No multiple physical instances, only multiple containers. 

I'd be content with simply being able to pass in a few env vars to the container:
- mode=restore-only
- s3 backup config to pull the backup
- I'm not sure if anything else is really necessary, everything else would be handled by the deployment pipeline

The hostname of the db container would be set by the deployment pipeline through an ENV var, and `docker compose up` would simply bind to that name. 

Thank you for the quick responses, I hope I explained the idea clearly enough.

**Comment by pellepelster** [comment]
I think I slowly see where this is going :-)

As per https://github.com/pellepelster/solidblocks/commit/a845eca83458dc3d4776b350281c1634f6bff437 there is now a read-only mode, that can be used as a command when the DB is started (analogous to the maintenance mode here https://pellepelster.github.io/solidblocks/rds/usage/index.html#maintenance)

It is not released yet because it still need to be integrated with the TF module and I want to make sure that no accidental data loss can occur. I am leaning towards a solution to only provide read-only users when restore is active, but this would not work in your scenario right? (docker image is already available as ghcr.io/pellepelster/solidblocks-rds-postgresql:14-snapshot-rc)

**Comment by Levara** [comment]
Amazing. I'll have to test it out once I finish with work (tonight probably), but just from reading the code it seems to me that this is exactly what I needed. 

So I'd start the db with 
```
 docker run \
    --name instance1 \
    -e DB_INSTANCE_NAME=instance1 \
    -e DB_DATABASE_db1=database1 \
    -e DB_USERNAME_db1=user1 \
    -e DB_PASSWORD_db1=password1 \
    -e DB_BACKUP_S3=1 \
    -e DB_S3_OTHER_ENVS
    -v "$(pwd)/postgres_backup:/storage/backup" \
    -v "$(pwd)/postgres_data:/storage/data" \
    ghcr.io/pellepelster/solidblocks-rds-postgresql:14-v0.2.3 \
    restore-only
```
and it would pull the backup and configure the db. 

>  I am leaning towards a solution to only provide read-only users when restore is active, but this would not work in your scenario right?

No, in my intended scenario the users should be able to use the database normally, only the backups should be disabled (or configured to local only). However, we'd still have `rds` user as an admin right? In that case it wouldn't be such a big inconvenience to run a simple `ALTER ROLE` script on the db during `docker-compose up`. Ideally, we'd be able to simply pass in another ENV var or a switch to the command (`restore-only --write=true` or similar) to get it writable.

**Comment by pellepelster** [comment]
did this work out for you?

**Comment by Levara** [comment]
Sorry, it got super crazy super fast :)
I've finally managed to find the time to test everything out. So far, the only problem I've found is that starting the image with `restore-only` command does not set the `DB_MODE=restore-only` env variable. Everything works correctly if you also pass in the `DB_MODE` env variable to docker. Without it, it performs a WAL push when shutting down the container and pollutes the backup. But if I explicitly add it to the docker command, everything works beautifully. 
Thanks man, this is really useful.
